src/main/java/core/basesyntax/strategy/PurchaseOperationHandler.java

```diff
@@ -0,0 +1,28 @@
+package core.basesyntax.strategy;
+
+import core.basesyntax.db.Storage;
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.service.OperationHandler;
+import core.basesyntax.serviceimpl.FruitTransactionValidatorImpl;
+import java.util.Optional;
+
+public class PurchaseOperationHandler implements OperationHandler {
+    private FruitTransactionValidatorImpl validator;
+
+    @Override
+    public void apply(FruitTransaction fruitTransaction) {
+        validator.validateOperation(fruitTransaction);
+        String fruit = fruitTransaction.getFruit();
+        validator.validateFruit(fruit);
+        int amountToPurchase = fruitTransaction.getQuantity();
+        validator.validateAmount(amountToPurchase, FruitTransaction.Operation.PURCHASE);
+
+        Optional<Integer> currentBalance = Optional.ofNullable(Storage.fruitStorage.get(fruit));
+        int balanceAfterPurchase = currentBalance.orElse(0) - amountToPurchase;
+
+        if (balanceAfterPurchase < 0) {
+            throw new RuntimeException("Insufficient stock for item " + fruit);
+        }
+        Storage.fruitStorage.put(fruit, balanceAfterPurchase);
+    }
```

Comment:
The class `PurchaseOperationHandler` should be placed in its own package according to the single responsibility principle and project structure guidelines. Consider moving it to a package like `core.basesyntax.strategy.impl` to follow the convention of placing implementation classes in a separate package.